### PR TITLE
Hooks/feature

### PR DESCRIPTION
**Pull request by MaxDaniLOVE**
### Implemented:
- Move auth logic to Redux;
- Add HOC for pages with modal window;
- Add alerts for incorrect validation;
- Work on styles;
- Add hooks
- [Demo](https://dims-beta.netlify.app/)

### CODE REVIEW

File: src/hoc/withModal.js [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    static getDerivedStateFromProps(nextProps) {
+      const { members, tasks, progress } = nextProps;
+      const mainData = {
+        MEMBERS_PAGE: members,
+        TASK_PAGE: tasks,
+        TRACK_PAGE: progress,
+      };
+      const pageData = mainData[pageType];
```
Comment (maximsan) [inline src/hoc/withModal.js:106]: Let's move the code to helper method

File: src/hoc/withModal.js [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    componentDidMount() {
+      const {
+        match,
+        user: { userId },
+      } = this.props;
+      const {
+        params: { tid },
+      } = match;
```
Comment (maximsan) [inline src/hoc/withModal.js:95]: Do spread in one place

File: src/hoc/withModal.js [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+        addUserProgress,
+      } = props;
+      switch (pageType) {
+        case 'MEMBERS_PAGE':
+          this.getData = getUsers;
+          this.addData = addUser;
+          this.editData = editUser;
+          this.deleteData = deleteUser;
+          this.defaultInputsData = defaultRegisterData;
+          this.dataInputs = membersInputs;
```
Comment (maximsan) [inline src/hoc/withModal.js:60]: Let's create helper for this and remove repeating

File: src/hoc/withModal.js [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+      });
+    };
+
+    onDataOpen = (recievedId) => {
+      const { setFormData } = this.props;
+      const { pageData } = this.state;
+      const editedUser =
+        pageType === 'MEMBERS_PAGE'
+          ? pageData.find(({ id }) => id === recievedId)
+          : pageData.find(({ taskTrackId }) => taskTrackId === recievedId);
```
Comment (maximsan) [inline src/hoc/withModal.js:211]: create method for find user

File: src/hoc/withModal.js [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+          trackDate: newSubtask.trackDate,
+        };
+        const isFormValid = validation(validatedInputs, subtasksInputs);
+        setFormData(newSubtask);
+        this.setState({ isFormValid });
+      } else {
+        const validatedInputs = { ...updated };
+        const isFormValid = validation(validatedInputs, this.dataInputs);
+        setFormData(updated);
+        this.setState({ isFormValid });
```
Comment (maximsan) [inline src/hoc/withModal.js:172]: you can call this.setState once

File: src/hoc/withModal.js [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+          userName,
+          ...updated,
+        };
+
+        const validatedInputs = {
+          trackNote: newSubtask.trackNote,
+          trackDate: newSubtask.trackDate,
+        };
+        const isFormValid = validation(validatedInputs, subtasksInputs);
+        setFormData(newSubtask);
```
Comment (maximsan) [inline src/hoc/withModal.js:166]: It seems, you can create method for 3 lines of code and reuse below it too

File: src/hoc/withModal.js [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+
+    onSubtaskModalOpen = (taskId) => {
+      const { progress, setFormData, formData } = this.props;
+      const editedTask = progress.find(({ taskId: id }) => id === taskId);
+      const { taskName } = editedTask;
+      setFormData({ ...formData, taskId, taskName });
+      this.onModalOpen();
+      this.setState({
+        showModal: true,
+      });
```
Comment (maximsan) [inline src/hoc/withModal.js:135]: Why do you do the same twice ?

File: src/pages/AuthPage.js [hunk 8]
```diff
@@ -60,5 +60,12 @@ class AuthPage extends Component {
...
 }
-AuthPage.contextType = AuthContext;
-export default AuthPage;
+
+AuthPage.propTypes = {
+  registerUser: PropTypes.func.isRequired,
+  logIn: PropTypes.func.isRequired,
+};
+
+const mapDispatchToProps = (dispatch) => bindActionCreators({ logIn, registerUser }, dispatch);
```
Comment (maximsan) [inline src/pages/AuthPage.js:69]: it's better to wrap in brackets and return explicitly everywhere for support

File: src/pages/MembersPage.js [hunk 9]
```diff
@@ -1,199 +1,114 @@
...
+  onEditDataModalOpen,
+  onDataOpen,
+  onModalClose,
+  onModalOpen,
+  setFormData,
+}) => {
+  useEffect(() => {
+    setFormData(defaultRegisterData);
+  }, [setFormData]);
+  const modalHeader = isEditMode || isDetailMode ? <h3>User&apos;s details:</h3> : <h3>Add new user:</h3>;
```
Comment (maximsan) [inline src/pages/MembersPage.js:39]: wrap condition in brackets

File: src/store/actions/dataActions.js [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+        type: EDIT_TASK,
+      });
+      dispatch(getTasks());
+    } catch (error) {
+      errorCallback(dispatch, error);
+    }
+  };
+};
+
+const startFetchingData = () => ({ type: FETCH_DATA_START });
```
Comment (maximsan) [inline src/store/actions/dataActions.js:199]: Wrap in brackets and return explicitly

File: src/store/reducers/authReducer.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+const initialState = {
+  isLoggedIn: false,
+  user: {},
+};
+
+const authReducer = (state = initialState, { type, payload }) => {
+  switch (type) {
+    case AUTH_LOG_IN:
+    case AUTH_LOG_OUT:
+      return { ...state };
```
Comment (maximsan) [inline src/store/reducers/authReducer.js:12]: I guess we can return just state

File: src/utils/transformEditData.js [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    const { deadlineDate, startDate } = editedData;
+    return { ...editedData, deadlineDate: dateToString(deadlineDate), startDate: dateToString(startDate) };
+  }
+  if (pageType === 'TRACK_PAGE') {
+    const editedData = pageData.find(({ taskTrackId }) => taskTrackId === recievedId);
+    const { trackDate } = editedData;
+    return { ...editedData, trackDate: dateToString(trackDate) };
+  }
+  return null;
+};
```
Comment (maximsan) [inline src/utils/transformEditData.js:20]: Do you have unit tests ?

File: src/utils/noteConverter.js [hunk 13]
```diff
@@ -1,6 +1,11 @@
...
-  if (note.length >= 20) {
-    return `${note.slice(0, 15)}...`;
+const noteConverter = (note, length) => {
+  if (note.length > length) {
+    const substring = note.slice(0, length);
+    const lastSpaceIndex = substring.lastIndexOf(' ');
+    if (lastSpaceIndex < 0) {
+      return `${substring}...`;
+    }
+    return `${note.slice(0, lastSpaceIndex)}...`;
```
Comment (maximsan) [inline src/utils/noteConverter.js:8]: Do you have unit tests ?

File: src/hoc/withModal.js (hunk 2 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:95]: done

File: src/hoc/withModal.js (hunk 5 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:172]: done

File: src/hoc/withModal.js (hunk 7 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:135]: fixed

File: src/store/reducers/authReducer.js (hunk 11 above)
Comment (MaxDaniLOVE) [reply src/store/reducers/authReducer.js:12]: done

File: src/hoc/withModal.js (hunk 1 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:106]: done

File: src/hoc/withModal.js (hunk 4 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:211]: done

File: src/hoc/withModal.js (hunk 3 above)
Comment (MaxDaniLOVE) [reply src/hoc/withModal.js:60]: done

File: src/utils/noteConverter.js (hunk 13 above)
Comment (MaxDaniLOVE) [reply src/utils/noteConverter.js:8]: fix previous tests

File: src/utils/transformEditData.js (hunk 12 above)
Comment (MaxDaniLOVE) [reply src/utils/transformEditData.js:20]: add test

File: src/store/actions/dataActions.js (hunk 10 above)
Comment (MaxDaniLOVE) [reply src/store/actions/dataActions.js:199]: fix

File: src/pages/MembersPage.js (hunk 9 above)
Comment (MaxDaniLOVE) [reply src/pages/MembersPage.js:39]: fix

File: src/pages/AuthPage.js (hunk 8 above)
Comment (MaxDaniLOVE) [reply src/pages/AuthPage.js:69]: fixed

**Pull request merged**